pkg/capabilities/triggers/cron_trigger.go

```diff
@@ -0,0 +1,221 @@
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-co-op/gocron/v2"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+const cronTriggerID = "cron-trigger@1.0.0"
+
+var cronTriggerInfo = capabilities.MustNewCapabilityInfo(
+	cronTriggerID,
+	capabilities.CapabilityTypeTrigger,
+	"A trigger to schedule workflow execution to run periodically at fixed times, dates, or intervals.",
+)
+
+type cronTriggerConfig struct {
+	// The crontab to evaluate, with second granularity.
+	// The seconds field is required. The time zone will always be automatically set to UTC.
+	Schedule string `json:"schedule"`
+}
+
+type cronTriggerInput struct{}
+
+type CronTriggerMetadata struct{}
+
+type CronTriggerPayload struct {
+	ScheduledExecutionTime string
+}
+
+type CronTriggerResponse struct {
+	capabilities.TriggerEvent
+	Metadata CronTriggerMetadata
+	Payload  CronTriggerPayload
+}
+
+type cronTrigger struct {
+	ch  chan<- capabilities.TriggerRegistrationResponse
+	job gocron.Job
+}
+
+type CronTriggerService struct {
+	capabilities.CapabilityInfo
+	capabilities.Validator[cronTriggerConfig, cronTriggerInput, capabilities.CapabilityResponse]
+	lggr      logger.Logger
+	mu        sync.Mutex
+	scheduler gocron.Scheduler
+	triggers  map[string]cronTrigger
+}
+
+var _ capabilities.TriggerCapability = (*CronTriggerService)(nil)
+var _ services.Service = &CronTriggerService{}
+
+func NewCronTriggerService(lggr logger.Logger) *CronTriggerService {
+	location, _ := time.LoadLocation("UTC")
+	s, err := gocron.NewScheduler(gocron.WithLocation(location))
+	if err != nil {
+		return nil
+	}
+
+	return &CronTriggerService{
+		CapabilityInfo: cronTriggerInfo,
+		Validator:      capabilities.NewValidator[cronTriggerConfig, cronTriggerInput, capabilities.CapabilityResponse](capabilities.ValidatorArgs{Info: cronTriggerInfo}),
+		triggers:       map[string]cronTrigger{},
+		lggr:           logger.Named(lggr, "CronTrigger"),
+		scheduler:      s,
+	}
+}
+
+// Register a new trigger
+// Can register triggers before the service is actively scheduling via .Start()
+func (cts *CronTriggerService) RegisterTrigger(ctx context.Context, req capabilities.TriggerRegistrationRequest) (<-chan capabilities.TriggerRegistrationResponse, error) {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	config, err := cts.ValidateConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	var job gocron.Job
+	callbackCh := make(chan capabilities.TriggerRegistrationResponse, defaultSendChannelBufferSize)
+
+	allowSeconds := true
+	jobDef := gocron.CronJob(config.Schedule, allowSeconds)
+
+	task := gocron.NewTask(func() {
+		// use scheduled time to generate a deterministic triggerEventId
+		scheduledExecutionTime, errr := job.LastRun()
```

Comment:
@justinkaseman I don't know if this will quite work for larger granularities of cron schedules; this is probably worth a test

For example: we need to verify that if we have a CRON schedule of every 5 minutes that the scheduled execution time will be 00:05:00/00:10:00/00:15:00 for all nodes, and not:

00:05:01 for one node
00:05:02 for another
and so on

Lmk if that makes sense?

Reply:
Makes sense. I'll add a long running test

Reply:
Added some tests, see if you think they sufficiently cover what you are talking about